### PR TITLE
Feature/neptune pete

### PR DESCRIPTION
**Pull request by peterhoward42**
### What

- Update the hard-coded response from neptune.GetCodeList to use a real database query.
- As the first illustrative use-case of upgrading the remainder of the CodeList implementation.
- Change the gremgo.Pool field in the neptune.driver.NeptuneDriver from being a struct type, to being
an interface.
- So that it is ready to be mocked for unit tests coming soon.

### How to review

Run this vendored version of dp-graph against the hierarchy builder and make sure nothing has been broken.

### Who can review

Describe who worked on the changes, so that other people can review = Pete H

### CODE REVIEW

File: DESIGN-NOTES.md [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+the **CodeList** API service has for a graph database. We'll use this as
+an example.
+
+## API
+
+The library API is defined by the graph and graph.driver packages.
+
+The **graph** package provides the **DB** type (a struct) - which is the
+object clients interact with. Clients interested in the Code List use-case
+can construct a DB specialised to this case with graph.NewCodeListStore().
```
Comment (gedge) [inline DESIGN-NOTES.md]:
```suggestion
can construct a DB specialised to this case with `graph.NewCodeListStore()`.
```

use code markup - typ.

File: DESIGN-NOTES.md [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+Other use-cases are catered for; like NewHierarchyStore().
```
Comment (gedge) [inline DESIGN-NOTES.md]:
```suggestion
Other use-cases are catered for; like `NewHierarchyStore()`.
```

File: DESIGN-NOTES.md [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+
+The API contract offered by NewCodeListStore() is defined by the
```
Comment (gedge) [inline DESIGN-NOTES.md]:
```suggestion
The API contract offered by `NewCodeListStore()` is defined by the 
```

File: neptune/driver/driver.go [hunk 4]
```diff
@@ -7,7 +7,7 @@ import (
 )
 
 type NeptuneDriver struct {
-	Pool *gremgo.Pool
+	Pool NeptunePool // Defined with an interface to support mocking.
```
Comment (daiLlew) [inline neptune/driver/driver.go]: 💯

File: DESIGN-NOTES.md [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+
+The library API is defined by the graph and graph.driver packages.
+
+The **graph** package provides the **DB** type (a struct) - which is the
+object clients interact with. Clients interested in the Code List use-case
+can construct a DB specialised to this case with graph.NewCodeListStore().
+Other use-cases are catered for; like NewHierarchyStore().
+
+The API contract offered by NewCodeListStore() is defined by the 
+graph.driver.CodeList interface. For example GetCodeList().
```
Comment (gedge) [inline DESIGN-NOTES.md]:
```suggestion
`graph.driver.CodeList` interface. For example `GetCodeList()`.
```

File: DESIGN-NOTES.md [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+
+Note that graph.DB uses Go's embedded fields. For example the graph.DB struct 
+contains an embedded driver.CodeList. It being "embedded" means that the
+methods of CodeList can be accessed either through the implicit attribute 
+name "CodeList", **or** the shortcut form of a **promoted** method like 
+this: DB.GetCodeList().
+
+Note also the way the config package sets all the DB attributes typified by
+DB.Codelist, either to nil or to the (typecast) config.Driver attribute.
+Which suggests these attributes differ, but they not! If the database variant
```
Comment (gedge) [inline DESIGN-NOTES.md]:
```suggestion
Which suggests these attributes differ, _but they do not_! If the database variant
```

File: DESIGN-NOTES.md [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+in use supports say the CodeList interface and the Hierarchy inteface, and
+both of these are selected, then both the CodeList attribute and the Hierarchy 
+attribute will be set - but with the same Neptune driver object.
+
+## Diverse Database Implementations
```
Comment (gedge) [inline DESIGN-NOTES.md]: this (more general explanation) section might be better moved above the more detailed section above - introducing the reader more gently

File: neptune/codelists.go [hunk 8]
```diff
@@ -45,7 +45,9 @@ func (n *NeptuneDB) GetCodeLists(ctx context.Context, filterBy string) (*models.
 
 // GetCodeList provides a CodeList for a given ID, having checked it exists
 // in the database. Nb. The caller is expected to fully qualify the embedded
-// Links field afterwards.
+// Links field afterwards. It returns an error if:
+// - The Gremlin query failed to execute.
+// - The requested CodeList does not exist.
```
Comment (gedge) [inline neptune/codelists.go]:
```suggestion
// - The requested CodeList does not exist (error is `ErrNotFound`).
```

File: neptune/driver/neptunepool.go [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+/*
+Interface NeptunePool defines the contract required of the gremgo 
+connection Pool by the Neptune.Driver.
+*/
+type NeptunePool interface {
+
+    Close()
+
+    Execute(query string, bindings, rebindings map[string]string) (
+        resp []gremgo.Response, err error)
```
Comment (gedge) [inline neptune/driver/neptunepool.go]: I find this line-break position very unusual and not easy to read - I'd much rather see each function in one-line.  Please consider one-liners.

typ.

File: DESIGN-NOTES.md [hunk 10]
```diff
@@ -0,0 +1,65 @@
+# Design Notes
```
Comment (gedge) [inline DESIGN-NOTES.md]: Worth putting a link to this doc from `README.md`

**Pull request merged**